packages/@aws-cdk/aws-ec2/lib/launch-template.ts

```diff
@@ -0,0 +1,707 @@
+
+  /**
+   * UserData executed by instances that are launched from this template.
+   * @internal
+   */
+  protected readonly _userData?: UserData;
+
+  // =============================================
+
+  constructor(scope: Construct, id: string, props: LaunchTemplateProps = {}) {
+    super(scope, id);
+
+    // Basic validation of the provided spot block duration
+    const spotDuration = props?.spotOptions?.blockDuration?.toHours({ integral: true });
+    if (spotDuration !== undefined && (spotDuration < 1 || spotDuration > 6)) {
+      // See: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-requests.html#fixed-duration-spot-instances
+      Annotations.of(this).addError('Spot block duration must be exactly 1, 2, 3, 4, 5, or 6 hours.');
+    }
+
+    this._role = props.role;
+    this._grantPrincipal = this._role;
+    const iamProfile: iam.CfnInstanceProfile | undefined = this._role ? new iam.CfnInstanceProfile(this, 'Profile', {
+      roles: [this._role!.roleName],
+    }) : undefined;
+    const iamProfileToken = Lazy.string({
+      produce: () => {
+        if (iamProfile) {
+          return iamProfile.getAtt('Arn').toString();
+        }
+        return undefined;
+      },
+    });
+
+    if (props.securityGroup) {
+      this._connections = new Connections({ securityGroups: [props.securityGroup] });
+    }
+    const securityGroupsToken = Lazy.list({
+      produce: () => {
+        if (this._connections && this._connections.securityGroups.length > 0) {
+          return this._connections.securityGroups.map(sg => sg.securityGroupId);
+        }
+        return undefined;
+      },
+    });
+
+    if (props.userData) {
+      this._userData = props.userData;
+    }
+    const userDataToken = Lazy.string({
+      produce: () => {
+        if (this._userData) {
+          return Fn.base64(this._userData.render());
+        }
+        return undefined;
+      },
+    });
+
+    const imageConfig: MachineImageConfig | undefined = props.machineImage?.getImage(this);
+    if (imageConfig) {
+      this._osType = imageConfig.osType;
+    }
+
+    let marketOptions: any = undefined;
+    if (props?.spotOptions) {
+      marketOptions = {
+        marketType: 'spot',
+        spotOptions: {
+          blockDurationMinutes: spotDuration !== undefined ? spotDuration * 60 : undefined,
+          instanceInterruptionBehavior: props.spotOptions.interruptionBehavior,
+          maxPrice: props.spotOptions.maxPrice?.toString(),
+          spotInstanceType: props.spotOptions.requestType,
+          validUntil: props.spotOptions.validUntil?.date.toUTCString(),
+        },
+      };
+      // Remove SpotOptions if there are none.
+      if (Object.keys(marketOptions.spotOptions).filter(k => marketOptions.spotOptions[k]).length == 0) {
+        marketOptions.spotOptions = undefined;
+      }
+    }
+
+    this.tags = new TagManager(TagType.KEY_VALUE, 'AWS::EC2::LaunchTemplate');
+    const tagsToken = Lazy.any({
```

Comment:
Clever. Doesn't look too horrible. It may be unexpected that the L2 is taggable while the L1 isn't. But I can live with this, and if we have tests on it we can always refactor later.